relay-server/src/actors/store.rs

```diff
@@ -58,33 +58,143 @@ fn make_distinct_id(s: &str) -> Uuid {
+                    topic_name: topic_name.to_string(),
+                    producer: Arc::clone(producer),
+                });
+            }
+
+            for config_p in params {
+                client_config.set(config_p.name.as_str(), config_p.value.as_str());
+            }
 
-    if let Some(producer) = reused_producers.get(&config_name) {
-        return Ok(Arc::clone(producer));
+            let producer = Arc::new(
+                client_config
+                    .create_with_context(CaptureErrorContext)
+                    .context(ServerErrorKind::KafkaError)?,
+            );
+
+            reused_producers.insert(config_name, Arc::clone(&producer));
+            Ok(Producer::Single {
+                topic_name: topic_name.to_string(),
+                producer,
+            })
+        }
+        KafkaConfig::Sharded { shards, configs } => {
+            let mut producers = BTreeMap::new();
+            for (shard, kafka_params) in configs {
+                if let Some(producer) = reused_producers.get(&kafka_params.config_name) {
+                    let cached_producer = Arc::clone(producer);
+                    producers.insert(
+                        shard,
+                        (kafka_params.topic_name.to_string(), cached_producer),
+                    );
+                    continue;
+                }
+                for config_p in kafka_params.params {
+                    client_config.set(config_p.name.as_str(), config_p.value.as_str());
+                }
+                let producer = Arc::new(
+                    client_config
+                        .create_with_context(CaptureErrorContext)
+                        .context(ServerErrorKind::KafkaError)?,
+                );
+                reused_producers.insert(kafka_params.config_name, Arc::clone(&producer));
+                producers.insert(shard, (kafka_params.topic_name.to_string(), producer));
+            }
+            Ok(Producer::Sharded(Sharded { shards, producers }))
+        }
     }
+}
 
-    let mut client_config = ClientConfig::new();
+/// This object containes the Kafka producer variants for single and sharded configurations.
+pub enum Producer {
+    Single {
+        topic_name: String,
+        producer: Arc<ThreadedProducer>,
+    },
+    Sharded(Sharded),
+}
 
-    for config_p in kafka_config {
-        client_config.set(config_p.name.as_str(), config_p.value.as_str());
+/// Sharded producer configuration.
+pub struct Sharded {
+    /// The maximum number of shards for this producer.
+    shards: u64,
+    /// The actual Kafka producer assigned to the range of logical shards, where the `u64` in the map is
+    /// the inclusive beginning of the range.
+    producers: BTreeMap<u64, (String, Arc<ThreadedProducer>)>,
+}
+
+impl Sharded {
+    /// Returns topic name and the Kafka producer based on the provided sharding key.
+    pub fn get_producer(&self, sharding_key: u64) -> (&str, Arc<ThreadedProducer>) {
```

Comment:
Might be a premature generalization, but we could make this function generic:
```suggestion
    pub fn get_producer(&self, sharding_key: impl Hash) -> (&str, Arc<ThreadedProducer>) {
```